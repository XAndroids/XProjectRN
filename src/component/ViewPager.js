import React, {Component} from 'react';
import {StyleSheet, View, ViewPagerAndroid, ScrollView, Platform} from "react-native";

/**
 * 根据android和ios系统的差异，分别使用RN的ViewPagerAndroid和ScrollView实现了Android ViewPager的组件效果
 */
export default class ViewPager extends Component {
    //声明并初始化State
    state = {
        width: 0,
        height: 0,
        initialIndex: this.props.initialIndex,
    };

    constructor(props: Props) {
        super(props);
        //声明实例属性，并且绑定事件
        this.adjustCardSize = this.adjustCardSize.bind(this);
    }

    render() {
        //根据平台差异，使用ViewPagerAndroid和ScrollView来实现
        if (Platform.OS === 'android') {
            return this.renderAndroid();
        } else {
            return this.renderIOS();
        }
    }

    renderAndroid() {
        return (
            //使用ViewPagerAndroid在Android平台渲染
            <ViewPagerAndroid
                initialPage={this.state.initialIndex}
                style={[styles.viewpager, this.props.style]}>
                {this.renderContent()}
            </ViewPagerAndroid>
        );
    }

    renderIOS() {
        return (
            //使用ScrollView在iOS平台上渲染
            <ScrollView
                style={[styles.viewpager, this.props.style]}
                contentOffset={{x: this.state.width * this.state.initialIndex, y: 0}}
                horizontal={true}
                showsHorizontalScrollIndicator={false}
                showsVerticalScrollIndicator={false}
                pagingEnabled={true}
                onLayout={this.adjustCardSize}>
                {this.renderContent()}
            </ScrollView>
        );
    }

    adjustCardSize(e) {
        //当布局改变的时候，重新获取ViewPager组件的宽高，使得子内容组件填充ViewPager组件大小
        this.setState({
            width: e.nativeEvent.layout.width,
            height: e.nativeEvent.layout.height,
        });
    }

    renderContent() {
        const {width, height} = this.state;
        return React.Children.map(this.props.children, (child, i) => (
            <View style={{width, height}}>
                {child}
            </View>
        ));
    }
}


var styles = StyleSheet.create({
    viewpager: {
        flex: 1,
    },
});
