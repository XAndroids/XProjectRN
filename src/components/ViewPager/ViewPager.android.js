import React, { Component } from 'react'
import { ViewPagerAndroid, View } from 'react-native'
import PropTypes from 'prop-types'
import { styles } from './styles'

export default class ViewPager extends Component {
    render() {
        return (
            //使用ViewPagerAndroid在Android平台渲染
            <ViewPagerAndroid
                ref={this.props.viewPagerRef}
                initialPage={this.props.initPageIndex}
                style={[styles.viewpager, this.props.style]}
                onPageSelected={this.handlerPagerSwitch.bind(this)}>
                {this.renderContent()}
            </ViewPagerAndroid>
        );
    }

    renderContent(width, height) {
        return React.Children.map(this.props.children, (child, i) => {
            return (
                <View>
                    {child}
                </View>
            );
        });
    }

    handlerPagerSwitch(event) {
        this.props.onPageSwitch(event.nativeEvent.position);
    }
}

//FIXME 如何在不同平台复用同样的propTypes、defaultProps和styles
ViewPager.propTypes = {
    initPageIndex: PropTypes.number,
    onPageSwitch: PropTypes.func
};

ViewPager.defaultProps = {
    initPageIndex: 0
};
