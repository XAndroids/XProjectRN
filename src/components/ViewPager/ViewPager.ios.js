import React, {Component} from 'react'
import {ScrollView} from "react-native"
import PropTypes from 'prop-types'
import {getWindowsWidth} from '../../utils'
import {styles} from './styles'

export default class ViewPager extends Component {
    render() {
        return (
            //使用ScrollView在iOS平台上渲染
            <ScrollView
                ref={this.props.viewPagerRef}
                style={[styles.viewpager, this.props.style]}
                contentOffset={{x: getWindowsWidth() * this.props.initPageIndex, y: 0}}
                horizontal={true}
                showsHorizontalScrollIndicator={false}
                pagingEnabled={true}
                onMomentumScrollEnd={this.handlerScrollSwitch.bind(this)}>
                {this.renderContent()}
            </ScrollView>
        );
    }

    renderContent(width, height) {
        return React.Children.map(this.props.children, (child, i) => {
            return child;
        });
    }

    handlerScrollSwitch(event) {
        this.props.onPageSwitch(event.nativeEvent.contentOffset.x / getWindowsWidth())
    }
}

//FIXME 如何在不同平台复用同样的propTypes、defaultProps和styles
ViewPager.propTypes = {
    initPageIndex: PropTypes.number,
    onPageSwitch: PropTypes.func
};

ViewPager.defaultProps = {
    initPageIndex: 0
};

