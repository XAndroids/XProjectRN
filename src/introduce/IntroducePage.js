import React, { Component } from 'react'
import { View, Text, Image, StyleSheet, Platform } from 'react-native';
import PropTypes from 'prop-types';
import { getWindowsWidth, getWindowsHeight } from '../utils/Devices';

/**
 * 定义了Introduce页面ViewPager子页面的内容
 */
export default class IntroducePage extends Component {

    render() {
        return (
            <View style={[styles.page, {
                backgroundColor: this.props.backgroundColor,
                width: getWindowsWidth(),
                height: getWindowsHeight()
            }]}>
                {this.renderText(this.props.title, 28)}
                <Image style={styles.image} source={require('../../img/introduce_icon1.png')} />
                {this.renderText(this.props.content, 16)}
            </View>
        )
    }

    /**
   * 渲染文案，iOS平台Text无法通过textAlignVertical实现字体居中，则使用View容器包裹实现
   * @param {显示的文案} text
   * @param {显示的文案字体大小} fontSize
   */
    renderText(text, fontSize) {
        if (Platform.OS === 'ios') {
            return (
                <View style={{ flex: 1, justifyContent: 'center' }}>
                    <Text style={[styles.text, { fontSize: fontSize }]}>{text}</Text>
                </View>);
        } else {
            return (
                <Text style={[styles.text, { fontSize: fontSize }]}>{text}</Text>
            );
        }
    }
}

IntroducePage.propTypes = {
    title: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
};

const styles = StyleSheet.create({
    page: {
        flex: 1,
        alignItems: 'center'
    },
    text: {
        color: 'white',
        ...Platform.select({
            android: {
                flex: 1,
                textAlign: 'center',
                textAlignVertical: 'center',
            },
        }),
    },
    image: {
        flex: 1,
    },
});

