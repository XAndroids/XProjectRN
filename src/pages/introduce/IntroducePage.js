import React, {Component} from 'react'
import {View, Text, Image, StyleSheet, Platform} from 'react-native'
import PropTypes from 'prop-types'
import {getWindowsWidth, getWindowsHeight} from '../../utils'

/**
 * 定义了Introduce页面ViewPager子页面的内容
 */
export default class IntroducePage extends Component {

    render() {
        const {backgroundColor, title, content, image} = this.props;

        return (
            <View style={[styles.page, {
                backgroundColor: backgroundColor,
                width: getWindowsWidth(),
                height: getWindowsHeight()
            }]}>
                {this._renderText(title, 28, 'center', 'center')}
                <Image style={{flex: 1}} source={image}/>
                {this._renderText(content, 16, 'flex-end', 'bottom')}
                <View style={{
                    backgroundColor: 'rgba(0, 0, 0, 0.1)',
                    height: 1,
                    width: getWindowsWidth(),
                    marginBottom: 50,
                    marginTop: 20
                }}/>
            </View>
        )
    }

    _renderText(text, fontSize, justifyContent, textAlignVertical) {
        if (Platform.OS === 'ios') {
            return (
                //渲染文案，iOS平台Text无法通过textAlignVertical实现字体居中，则使用View容器包裹实现
                <View style={{flex: 1, justifyContent: justifyContent}}>
                    <Text style={[styles.text, {fontSize: fontSize}]}>{text}</Text>
                </View>);
        } else {
            return (
                <Text style={[styles.text, {
                    fontSize: fontSize,
                    textAlignVertical: textAlignVertical
                }]}>{text}</Text>
            );
        }
    }
}

IntroducePage.propTypes = {
    title: PropTypes.string.isRequired,
    image: PropTypes.number.isRequired,
    content: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
};

const styles = StyleSheet.create({
    page: {
        flex: 1,
        alignItems: 'center'
    },
    text: {
        color: 'white',
        textAlign: 'center',
        ...Platform.select({
            android: {
                flex: 1
            },
        }),
    }
});

