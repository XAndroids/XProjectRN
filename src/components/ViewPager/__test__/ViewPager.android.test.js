import {fireEvent, render, shallow} from 'react-native-testing-library'
import React from "react";
import {ViewPagerAndroid} from 'react-native'
import IntroducePage from "../../../pages/introduce/IntroducePage";
import ViewPager from "../../ViewPager/ViewPager.android";

const pages = [
    {
        title: 'Welcome',
        image: require('../../../../img/introduce_icon1.png'),
        content: 'Thank you for installing Paperboy.What makes Paperboy different form other news reader apps is its simplicity and elegant design.',
        backgroundColor: '#01A3AE',
    }, {
        title: 'Customizations',
        image: require('../../../../img/introduce_icon2.png'),
        content: 'We believe that each user is unique and hence several customization options are provided to suit different reading styles.To customize please visit the settings page.',
        backgroundColor: '#4CAF50',
    }, {
        title: 'Reading pattern learner',
        image: require('../../../../img/introduce_icon3.png'),
        content: 'The home screen tiles will automatically reaarrange based on reading patterns for better user experence.All data is stored locally kepping in mind user privacy.',
        backgroundColor: '#673AB8',
    }
];

describe("ViewPager.android property test", () => {
    it("initialPage property test", () => {
        const initPageIndex = 1;
        const {getByType} = render(<ViewPager initPageIndex={initPageIndex}>
            {pages.map((page, i) => <IntroducePage
                key={i}
                title={page.title}
                image={page.image}
                content={page.content}
                backgroundColor={page.backgroundColor}/>)}
        </ViewPager>);

        expect(getByType(ViewPagerAndroid).props.initialPage).toBe(initPageIndex);
    })

    it("child property test", () => {
        const initPageIndex = 1;
        const {getAllByType} = render(<ViewPager initPageIndex={initPageIndex}>
            {pages.map((page, i) => <IntroducePage
                key={i}
                title={page.title}
                image={page.image}
                content={page.content}
                backgroundColor={page.backgroundColor}/>)}
        </ViewPager>);
        const introduecePages = getAllByType(IntroducePage);
        expect(introduecePages.length).toBe(pages.length);
        for(let page_i = 0; page_i < introduecePages.length;page_i ++){
            expect(introduecePages[page_i].props.title).toBe(pages[page_i].title);
            expect(introduecePages[page_i].props.image).toBe(pages[page_i].image);
            expect(introduecePages[page_i].props.content).toBe(pages[page_i].content);
            expect(introduecePages[page_i].props.backgroundColor).toBe(pages[page_i].backgroundColor);
        }

    })
})

describe("ViewPager.android fun test", () => {
    it.skip("onPageSwitch fun test", () => {
        const initPageIndex = 1;
        const onPageSwitch = jest.fn();
        //FIXE mockResolvedValue Mock参数？？
        onPageSwitch.mockResolvedValue(1);

        const {getByType} = render(<ViewPager initPageIndex={initPageIndex} onPageSwitch={onPageSwitch}>
            {pages.map((page, i) => <IntroducePage
                key={i}
                title={page.title}
                image={page.image}
                content={page.content}
                backgroundColor={page.backgroundColor}/>)}
        </ViewPager>);

        //TypeError: Cannot read property 'nativeEvent' of undefined
        //FIXME 如何Mock掉方法参数:event.nativeEvent.position,this.props.onPageSwitch(event.nativeEvent.position);
        fireEvent(getByType(ViewPagerAndroid), "onPageSelected");
        expect(onPageSwitch).toHaveBeenCalled();
    })
})

describe("ViewPager.android Snapshot test",()=>{
    it("Snapshot test",()=>{
        const initPageIndex = 1;
        const onPageSwitch = jest.fn();
        const {toJSON} = render(<ViewPager initPageIndex={initPageIndex} onPageSwitch={onPageSwitch}>
            {pages.map((page, i) => <IntroducePage
                key={i}
                title={page.title}
                image={page.image}
                content={page.content}
                backgroundColor={page.backgroundColor}/>)}
        </ViewPager>);

        expect(toJSON()).toMatchSnapshot();
    })
});

describe("ViewPager.android Shallow test",()=>{
    it("Shallow test",()=>{
        const initPageIndex = 1;
        const onPageSwitch = jest.fn();
        const {output} = shallow(<ViewPager initPageIndex={initPageIndex} onPageSwitch={onPageSwitch}>
            {pages.map((page, i) => <IntroducePage
                key={i}
                title={page.title}
                image={page.image}
                content={page.content}
                backgroundColor={page.backgroundColor}/>)}
        </ViewPager>);

        expect(output).toMatchSnapshot();
    })
});
