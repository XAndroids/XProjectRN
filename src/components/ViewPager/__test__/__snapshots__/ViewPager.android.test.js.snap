// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewPager.android Shallow test Shallow test 1`] = `
<UnimplementedView
  initialPage={1}
  onPageSelected={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component>
    <IntroducePage
      backgroundColor="#01A3AE"
      content="Thank you for installing Paperboy.What makes Paperboy different form other news reader apps is its simplicity and elegant design."
      image={
        Object {
          "testUri": "../../../img/introduce_icon1.png",
        }
      }
      title="Welcome"
    />
  </Component>
  <Component>
    <IntroducePage
      backgroundColor="#4CAF50"
      content="We believe that each user is unique and hence several customization options are provided to suit different reading styles.To customize please visit the settings page."
      image={
        Object {
          "testUri": "../../../img/introduce_icon2.png",
        }
      }
      title="Customizations"
    />
  </Component>
  <Component>
    <IntroducePage
      backgroundColor="#673AB8"
      content="The home screen tiles will automatically reaarrange based on reading patterns for better user experence.All data is stored locally kepping in mind user privacy."
      image={
        Object {
          "testUri": "../../../img/introduce_icon3.png",
        }
      }
      title="Reading pattern learner"
    />
  </Component>
</UnimplementedView>
`;

exports[`ViewPager.android Snapshot test Snapshot test 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "flex-start",
        "borderColor": "red",
        "borderWidth": 1,
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
          },
          Object {
            "backgroundColor": "#01A3AE",
            "height": 1334,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "textAlign": "center",
              },
              Object {
                "fontSize": 28,
              },
            ]
          }
        >
          Welcome
        </Text>
      </View>
      <Image
        source={
          Object {
            "testUri": "../../../img/introduce_icon1.png",
          }
        }
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "textAlign": "center",
              },
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          Thank you for installing Paperboy.What makes Paperboy different form other news reader apps is its simplicity and elegant design.
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "height": 1,
            "marginBottom": 50,
            "marginTop": 20,
            "width": 750,
          }
        }
      />
    </View>
  </View>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
          },
          Object {
            "backgroundColor": "#4CAF50",
            "height": 1334,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "textAlign": "center",
              },
              Object {
                "fontSize": 28,
              },
            ]
          }
        >
          Customizations
        </Text>
      </View>
      <Image
        source={
          Object {
            "testUri": "../../../img/introduce_icon2.png",
          }
        }
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "textAlign": "center",
              },
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          We believe that each user is unique and hence several customization options are provided to suit different reading styles.To customize please visit the settings page.
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "height": 1,
            "marginBottom": 50,
            "marginTop": 20,
            "width": 750,
          }
        }
      />
    </View>
  </View>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
          },
          Object {
            "backgroundColor": "#673AB8",
            "height": 1334,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "textAlign": "center",
              },
              Object {
                "fontSize": 28,
              },
            ]
          }
        >
          Reading pattern learner
        </Text>
      </View>
      <Image
        source={
          Object {
            "testUri": "../../../img/introduce_icon3.png",
          }
        }
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "textAlign": "center",
              },
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          The home screen tiles will automatically reaarrange based on reading patterns for better user experence.All data is stored locally kepping in mind user privacy.
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "height": 1,
            "marginBottom": 50,
            "marginTop": 20,
            "width": 750,
          }
        }
      />
    </View>
  </View>
</View>
`;
